File: app/src/main/java/com/juanmacuevas/shoottheflakup/BulletsControl.java
package com.juanmacuevas.shoottheflakup;

import android.graphics.Canvas;

import java.util.ArrayList;
import java.util.Iterator;

class BulletsControl {
    private final float tankBottom;
    private ArrayList<FunctionalBullet> bullets;

    public BulletsControl(float tankBottom) {
        bullets = new ArrayList<>();
        this.tankBottom = tankBottom;
    }

    public void draw(Canvas canvas) {
        for(Iterator it = bullets.iterator(); it.hasNext();){
            FunctionalBullet b = (FunctionalBullet) it.next();
            if (b!=null)
                b.draw(canvas);
        }

    }

    public void update(long timer) {
        FunctionalBullet b;
        int i = 0;
        while (!bullets.isEmpty() && bullets.size() > i) {
            b = bullets.get(i);
            if (b.isOver())
                bullets.remove(i);
            else {
                b.update(timer);
                i++;
            }
        }
    }

    public void addBullet(int power, float angle, int x, int y) {
        bullets.add(new FunctionalBullet(power, angle,x,y,tankBottom));

    }

    public Iterable<FunctionalBullet> iterable() {
        return bullets;
    }
}

File: app/src/main/java/com/juanmacuevas/shoottheflakup/BulletsControl.kt
package com.juanmacuevas.shoottheflakup

import android.content.res.Resources
import android.graphics.Canvas
import android.util.DisplayMetrics
import androidx.core.util.Pair

import java.util.ArrayList

internal class BulletsControl(private val res: Resources, private val metrics: DisplayMetrics) {
    private val bullets = ArrayList<FunctionalBullet>()

    fun draw(canvas: Canvas) {
        val it = bullets.iterator()
        while (it.hasNext()) {
            val b = it.next()
            b.draw(canvas)
        }

    }

    fun update(timer: Long) {
        var b: FunctionalBullet
        var i = 0
        while (!bullets.isEmpty() && bullets.size > i) {
            b = bullets[i]
            if (b.isOver)
                bullets.removeAt(i)
            else {
                b.update(timer)
                i++
            }
        }
    }

    fun addBullet(power: Int, angle: Float, bulletOrigin: Pair<Int, Int>?) {
        bullets.add(FunctionalBullet(res, power, angle, bulletOrigin, metrics))

    }

    fun iterable(): Iterable<FunctionalBullet> {
        return bullets
    }
}
